src/S3/PostObjectV4.php

```diff
@@ -0,0 +1,209 @@
+<?php
+namespace Aws\S3;
+
+use Aws\Credentials\CredentialsInterface;
+use GuzzleHttp\Psr7\Uri;
+use Aws\Signature\SignatureTrait;
+use Aws\Common\Enum\DateFormat;
+use Aws\Api\TimestampShape as TimestampShape;
+
+/**
+ * Encapsulates the logic for getting the data for an S3 object POST upload form
+ *
+ * @link http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html
+ * @link http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-post-example.html
+ */
+class PostObjectV4
+{
+    use SignatureTrait;
+
+    const ISO8601_BASIC = 'Ymd\THis\Z';
+
+    private $ldt;
+    private $sdt;
+    private $client;
+    private $bucket;
+    private $formAttributes;
+    private $formInputs;
+    private $jsonPolicy;
+
+    /**
+     * Constructs the PostObject.
+     *
+     * The options array accepts the following keys:
+     * @link http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+     *
+     * @param S3ClientInterface $client     Client used with the POST object
+     * @param string            $bucket     Bucket to use
+     * @param array             $formInputs Associative array of form input
+     *                                      fields.
+     * @param array             $options    Policy condition options
+     * @param mixed             $expiration Upload expiration time value. By
+     *                                      default: 1 hour vaild peroid.
+     */
+    public function __construct(
+        S3ClientInterface $client,
+        $bucket,
+        array $formInputs,
+        array $options = [],
+        $expiration = NULL
+    ) {
+        $this->ldt = gmdate(self::ISO8601_BASIC);
+        $this->sdt = substr($this->ldt, 0, 8);
+
+        $this->client = $client;
+        $this->bucket = $bucket;
+
+        // setup form attributes
+        $this->formAttributes = [
+            'action'  => $this->generateUri(),
+            'method'  => 'POST',
+            'enctype' => 'multipart/form-data'
+        ];
+
+        // setup basic policy
+        $expiration = is_null($expiration)
+                ? gmdate('Y-m-d\TH:i:s\Z', strtotime('+1 hours'))
+                : TimestampShape::format($expiration, 'iso8601');
+
+        $policy = [
+            'expiration' => $expiration,
+            'conditions' => [],
+        ];
+        foreach ($options as $option) {
+            $policy['conditions'][] = $option;
+        }
+
+        $this->formInputs = $formInputs + ['key' => '${filename}'];
+        // add v4 conditions for policy and formInputs
+        foreach ($this->getV4Conditions() as $key => $value) {
```

Comment:
There's a race condition here. You're resolving credentials in `getV4Conditions` and later on line 87. Credentials can be refreshable, which means that you might end up creating a policy with one set of credentials and then signing it with another. Resolve the credentials once.